gcloud-java-contrib/gcloud-java-nio/src/main/java/com/google/gcloud/storage/contrib/nio/CloudStorageFileSystemProvider.java

```diff
@@ -307,32 +356,40 @@ public void copy(Path source, Path target, CopyOption... options) throws IOExcep
     if (toPath.seemsLikeADirectoryAndUsePseudoDirectories()) {
       throw new CloudStoragePseudoDirectoryException(toPath);
     }
-    GcsFilename from = fromPath.getGcsFilename();
-    GcsFilename to = toPath.getGcsFilename();
-    GcsFileMetadata metadata = gcsService.getMetadata(from);
-    if (metadata == null) {
-      throw new NoSuchFileException(source.toString());
-    }
-    if (fromPath.equals(toPath)) {
-      return;
-    }
-    if (!wantReplaceExisting && gcsService.getMetadata(to) != null) {
-      throw new FileAlreadyExistsException(target.toString());
-    }
-    GcsFileOptions.Builder builder = wantCopyAttributes
-        ? copyFileOptions(metadata.getOptions())
-        : new GcsFileOptions.Builder();
-    GcsFileOptions fileOptions = buildFileOptions(builder, options);
-    try (GcsInputChannel input = gcsService.openReadChannel(from, 0);
-        GcsOutputChannel output = gcsService.createOrReplace(to, fileOptions)) {
-      ByteBuffer block = ByteBuffer.allocate(blockSize);
-      while (input.read(block) != -1) {
-        block.flip();
-        while (block.hasRemaining()) {
-          output.write(block);
+
+    try {
+
+
+      if (wantCopyAttributes) {
+        BlobInfo blobInfo = storage.get(fromPath.getBlobId());
+        if (!setCacheControl) {
+          tgtInfoBuilder.cacheControl(blobInfo.cacheControl());
         }
-        block.clear();
+        if (!setContentType) {
+          tgtInfoBuilder.contentType(blobInfo.contentType());
+        }
+        tgtInfoBuilder.acl(blobInfo.acl());
+        tgtInfoBuilder.metadata(blobInfo.metadata());
+      }
+
+      BlobInfo tgtInfo = tgtInfoBuilder.build();
+      Storage.CopyRequest.Builder copyReqBuilder = Storage.CopyRequest.builder()
+          .source(fromPath.getBlobId());
+      if (wantReplaceExisting) {
+        copyReqBuilder = copyReqBuilder.target(tgtInfo);
+      } else {
+        copyReqBuilder = copyReqBuilder.target(tgtInfo, Storage.BlobTargetOption.doesNotExist());
+      }
+      CopyWriter copyWriter = storage.copy(copyReqBuilder.build());
+      copyWriter.result();
+    } catch (StorageException oops) {
+      // RPC API can only throw StorageException, but CloudStorageFileSystemProvider
+      // can only throw IOException. So we tunnel the latter inside the former.
+      Throwables.propagateIfInstanceOf(oops.getCause(), IOException.class);
```

Comment:
I tried that earlier, but it fails the test because it explicitly checks for NoSuchFileException amd FileAlreadyExistsException. I've hardcoded those in since you seem to prefer that solution. It's not pretty, but it works so far (I've also factored it out so we only need this code once).